cranelift/codegen/src/dominator_tree.rs

```diff
@@ -447,177 +482,55 @@ impl DominatorTree {
             self.nodes[block].idom = self.stree[idom].block;
         }
     }
-}
-
-/// Optional pre-order information that can be computed for a dominator tree.
-///
-/// This data structure is computed from a `DominatorTree` and provides:
-///
-/// - A forward traversable dominator tree through the `children()` iterator.
-/// - An ordering of blocks according to a dominator tree pre-order.
-/// - Constant time dominance checks at the block granularity.
-///
-/// The information in this auxiliary data structure is not easy to update when the control flow
-/// graph changes, which is why it is kept separate.
-pub struct DominatorTreePreorder {
-    nodes: SecondaryMap<Block, ExtraNode>,
-
-    // Scratch memory used by `compute_postorder()`.
-    stack: Vec<Block>,
-}
-
-#[derive(Default, Clone)]
-struct ExtraNode {
-    /// First child node in the domtree.
-    child: PackedOption<Block>,
-
-    /// Next sibling node in the domtree. This linked list is ordered according to the CFG RPO.
-    sibling: PackedOption<Block>,
-
-    /// Sequence number for this node in a pre-order traversal of the dominator tree.
-    /// Unreachable blocks have number 0, the entry block is 1.
-    pre_number: u32,
-
-    /// Maximum `pre_number` for the sub-tree of the dominator tree that is rooted at this node.
-    /// This is always >= `pre_number`.
-    pre_max: u32,
-}
-
-/// Creating and computing the dominator tree pre-order.
-impl DominatorTreePreorder {
-    /// Create a new blank `DominatorTreePreorder`.
-    pub fn new() -> Self {
-        Self {
-            nodes: SecondaryMap::new(),
-            stack: Vec::new(),
-        }
-    }
-
-    /// Recompute this data structure to match `domtree`.
-    pub fn compute(&mut self, domtree: &DominatorTree) {
-        self.nodes.clear();
 
+    /// Compute dominator tree preorder information.
+    /// This populates child/sibling links and preorder numbers for fast dominance checks.
+    fn compute_domtree_preorder(&mut self) {
         // Step 1: Populate the child and sibling links.
         //
         // By following the CFG post-order and pushing to the front of the lists, we make sure that
         // sibling lists are ordered according to the CFG reverse post-order.
-        for &block in domtree.cfg_postorder() {
-            if let Some(idom) = domtree.idom(block) {
+        let postorder: Vec<Block> = self.postorder.clone();
```

Comment:
Why clone the postorder vec here? It looks like it's borrowed only while mutating other fields (`nodes`, `dfs_worklist`) below so the borrow checker should be able to understand the lack of borrow conflict...

Reply:
Indeed I may have been overly cautious there